triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala

```
@@ -134,15 +136,28 @@ object TriggerServiceFixture {
       a <- testFn(uri, client, ledgerProxy)
     } yield a
 
-    fa.onComplete { _ =>
-      serviceF.foreach({ case (_, system) => system ! Stop })
-      ledgerF.foreach(_._1.close())
-      toxiproxyF.foreach(_._1.destroy)
+    fa.transformWith { ta =>
+      for {
```

What exactly went wrong with `Future.sequence`? As far as I understand `Future.sequence` is not fail fast so it shouldn’t stop anything from running. Does running these things in parallel cause issues?

---

Actually, I think I might have misread something and it is fail fast which probably explains the issue here.

---

I don't think `Future.sequence` is fail-fast. It's passed already-running Futures and AFAIK, there's no way to stop them.

---

So the issue was that we tried to stop things in parallel? I’m not quite sure why that’s a problem.

---

It is fail-fast in the sense that if one future fails, it completed, allowing the next step (the restart) to proceed despite something else's shutdown being in progress.

---

I could possibly restore the parallel shutdown, if the forced success of this chain is really the relevant piece, but would rather get this in as is first.

---

Makes sense, thanks for the explanation! I’m not worried about losing parallelism so totally fine to get this in and ignore this.